In [1]:
%run data_gather-Copy1.ipynb

{'_type': 'album',
 'api_path': '/albums/102298',
 'cover_art_thumbnail_url': 'https://images.genius.com/dc9148a9c2eb963ae0b5564ef2042412.300x300x1.jpg',
 'cover_art_url': 'https://images.genius.com/dc9148a9c2eb963ae0b5564ef2042412.1000x1000x1.jpg',
 'full_title': 'Euphoria Mourning by Chris Cornell',
 'id': 102298,
 'name': 'Euphoria Mourning',
 'name_with_artist': 'Euphoria Mourning (artist: Chris Cornell)',
 'primary_artist_names': 'Chris Cornell',
 'pusher_channel': 'album-102298',
 'release_date_components': {'year': 1999, 'month': 9, 'day': 21},
 'release_date_for_display': 'September 21, 1999',
 'updated_at': 1739842718,
 'url': 'https://genius.com/albums/Chris-cornell/Euphoria-mourning',
 'current_user_metadata': {'permissions': ['create_comment'],
                           'excluded_permissions': ['set_album_color_gradient',
                                                    'edit',
                                                    'destroy',
                              

In [2]:
%run data_cleaning-no-punctuation.ipynb

{'_type': 'album',
 'api_path': '/albums/102298',
 'cover_art_thumbnail_url': 'https://images.genius.com/dc9148a9c2eb963ae0b5564ef2042412.300x300x1.jpg',
 'cover_art_url': 'https://images.genius.com/dc9148a9c2eb963ae0b5564ef2042412.1000x1000x1.jpg',
 'full_title': 'Euphoria Mourning by Chris Cornell',
 'id': 102298,
 'name': 'Euphoria Mourning',
 'name_with_artist': 'Euphoria Mourning (artist: Chris Cornell)',
 'primary_artist_names': 'Chris Cornell',
 'pusher_channel': 'album-102298',
 'release_date_components': {'year': 1999, 'month': 9, 'day': 21},
 'release_date_for_display': 'September 21, 1999',
 'updated_at': 1739842718,
 'url': 'https://genius.com/albums/Chris-cornell/Euphoria-mourning',
 'current_user_metadata': {'permissions': ['create_comment'],
                           'excluded_permissions': ['set_album_color_gradient',
                                                    'edit',
                                                    'destroy',
                              

In [3]:
#https://pudding.cool/projects/vocabulary/index.html
#https://www.nltk.org/book/ch01.html#counting-vocabulary

In [4]:
#import nltk

In [5]:
#Test for one song 

text = chris_cornell_df_without_labels['processed_track_lyrics'].iloc[0].split()
def lexical_diversity(text):
    if len(text) == 0:
        return 0
    return len(set(text)) / len(text)
#https://www.nltk.org/book/ch01.html#counting-vocabulary 
#need to split because processed_track_lyrics joined the tokens created with spacy

lexical_diversity(text)

#0.26973684210526316
#about 27 of every 100 words is unique for the first song in 'processed_track_lyrics' column 

0.26973684210526316

In [6]:
#Loop for lexical diversity value for each song in 'processed_track_lyrics' column 

song_word_diversity_df = chris_cornell_df_without_labels[["album_name", "track_title", "processed_track_lyrics", "release_year", "track_lyrics"]].copy()

In [7]:
#Create new lists to add as new columns to new dataframe 

word_diversity_score = []
word_count_processed_lyrics = []
unique_text_count_processed_lyrics = []

text = song_word_diversity_df['processed_track_lyrics']

for text in song_word_diversity_df["processed_track_lyrics"]:
    tokens = text.split()
    def lexical_diversity(text):
        if len(text) == 0:
            return 0
        return len(set(text)) / len(text)
    word_diversity_score.append(lexical_diversity(tokens) * 100)
    word_count_processed_lyrics.append(len(tokens))
    unique_text_count_processed_lyrics.append(len(set(tokens)))

    
song_word_diversity_df["word_diversity_score"] = word_diversity_score
song_word_diversity_df["unique_text_count_processed_lyrics"] = unique_text_count_processed_lyrics
song_word_diversity_df["word_count_processed_lyrics"] = word_count_processed_lyrics

song_word_diversity_df = song_word_diversity_df[
    ["album_name", "release_year", "track_title", "word_diversity_score", "unique_text_count_processed_lyrics", "word_count_processed_lyrics", "processed_track_lyrics", "track_lyrics"]
]

In [8]:
song_word_diversity_df.reset_index(drop=True, inplace=True)

In [9]:
pd.options.display.max_rows = None
song_word_diversity_df.sort_values("word_diversity_score", ascending=False).head(10)

,album_name,release_year,track_title,word_diversity_score,unique_text_count_processed_lyrics,word_count_processed_lyrics,processed_track_lyrics,track_lyrics
35,Ultramega OK,1988,Circle of Power,74.074074,40,54,day finally come smell revenge sweet strong rekindle flames rage assholes they burn knees we live fear trembling words apology bullet list' x' accomplished deed ol' big badass circle power comin' ya not necessarily dumb beautiful beautiful it's circle power it's it's circle spinnin' circle power it's different names circle power it's circle power,"[Verse 1] The day will finally come and The smell of revenge, so sweet and strong Rekindle the flames of rage All those assholes, they burn on their knees [Verse 2] Well, here we live To see their fear and their trembling words of apology A bullet for every name on the list An 'X' for every accomplished deed Ol' big badass circle of power is comin' to get ya! [Guitar Solo and vocal ad-libs] [Interlude] Not necessarily dumb, but just, uh…beautiful Beautiful [Outro] Well, it's the circle of power And it's all, it's all the same Ah, the circle, spinnin' the circle of power And it's under different names Oh, the circle of power, it's all the same The circle of power"
29,Ultramega OK,1988,Flower,73.770492,45,61,her vain parade her veins seventeen eyes purple green treated like queen she was borrowed selfesteem she would do her dance painful masquerade spinning you her web her vain parade her uniform studded brass steel kissing napkin lipstick stains smearing sincerity her vain parade her veins break time crept her she's early gray her reflection looks concerned flowers hit her grave,"[Intro] Along her vain parade Along her veins Huah! [Verse 1] All of seventeen, eyes a purple green Treated like a queen, she was on borrowed self-esteem [Verse 2] She would do her dance, a painful masquerade Spinning you into her web along her vain parade [Verse 3] In her uniform of studded brass and steel Kissing napkin, lipstick stains, and smearing sincerity [Chorus] Along her vain parade Along her veins [Instrumental Break] Heyyyy-ya-la-la-hey-hey [Verse 4] Time crept up on her, she's in early gray Her reflection looks concerned as flowers hit her grave"
8,Euphoria Mourning,1999,Sweet Euphoria,71.111111,64,90,sweet euphoria heart you you lost grace hands that harmed you daze thousand yawns lost love it i'm truly gone morning i sleep fire open palm death jesus plastic armies wouldn't bring me sweet ether eyes i'm blind them aching hour time wilts like flowers sleeping landmine pillows tired angels save love lasting sweet euphoria heart you stole touched broken things you love stars light candles warms face i can't remember gone dandelions falling like falling like daydream mangos diving swallows save love save love save love lasting sweet euphoria,"[Verse] Sweet euphoria Mine is the heart you own You lost the grace of the hands that harmed you In the daze of a thousand yawns Lost my love, as it is I'm truly gone In your morning I will sleep Fire on an open palm Death for Jesus and plastic armies Wouldn't bring me back again Sweet as ether eyes I'm blind to them [Chorus] And in your aching hour Time wilts like flowers Sleeping on land-mine pillows Tired angels Save my love for the lasting one [Verse 2] Sweet euphoria Mine is heart you stole Touched and broken are the things you love Using stars to light your candles Warms my face but I can't remember yours [Chorus] Gone are your dandelions Falling like mine Falling like daydream mangos Diving swallows Save my love, save my love Save my love for the lasting one Sweet Euphoria"
33,Ultramega OK,1988,667,68.181818,15,22,kill them steal gold feed boy's killer feed young prepare it save you feed boy hates kill feed young feed young cool,"[Interlude] Kill them all Steal their gold Feed my boy's killer Feed the young, prepare it, save you Feed the boy, hates to kill Feed the young Feed the young [Outro] (Cool)"
17,Higher Truth,2015,Through the Window,65.6250

In [10]:
top_10_diverse_df = song_word_diversity_df.sort_values("word_diversity_score", ascending=False)[['track_title', "release_year", 'album_name', 'word_diversity_score', 'unique_text_count_processed_lyrics', 'word_count_processed_lyrics']].head(10).reset_index(drop=True).copy()

Question 5: What are the top 10 songs by word diversity? 

0       Circle of Power
1                Flower
2        Sweet Euphoria
3                   667
4    Through the Window
5              Circling
6             Attrition
7      Beyond the Wheel
8      Mood for Trouble
9         Follow My Way

In [11]:
top_10_diverse_df['word_diversity_score']=top_10_diverse_df['word_diversity_score'].round(0)
top_10_diverse_df

,track_title,release_year,album_name,word_diversity_score,unique_text_count_processed_lyrics,word_count_processed_lyrics
0,Circle of Power,1988,Ultramega OK,74.0,40,54
1,Flower,1988,Ultramega OK,74.0,45,61
2,Sweet Euphoria,1999,Euphoria Mourning,71.0,64,90
3,667,1988,Ultramega OK,68.0,15,22
4,Through the Window,2015,Higher Truth,66.0,105,160
5,Circling,2015,Higher Truth,65.0,81,124
6,Attrition,2012,King Animal,65.0,57,88
7,Beyond the Wheel,1988,Ultramega OK,63.0,45,71
8,Mood for Trouble,1988,Ultramega OK,60.0,33,55
9,Follow My Way,1999,Euphoria Mourning,60.0,78,130


In [12]:
scatter_song_word_diversity_df=song_word_diversity_df.copy()

In [13]:
scatter_song_word_diversity_df = scatter_song_word_diversity_df.drop(columns = ['track_lyrics', 'processed_track_lyrics', "unique_text_count_processed_lyrics", "word_count_processed_lyrics"])

In [14]:
scatter_song_word_diversity_df=scatter_song_word_diversity_df[['track_title', "album_name", 'release_year', 'word_diversity_score']]

In [15]:
scatter_song_word_diversity_df

,track_title,album_name,release_year,word_diversity_score
0,Can’t Change Me,Euphoria Mourning,1999,26.973684
1,Flutter Girl,Euphoria Mourning,1999,39.230769
2,Preaching the End of the World,Euphoria Mourning,1999,39.306358
3,Follow My Way,Euphoria Mourning,1999,60.000000
4,When I’m Down,Euphoria Mourning,1999,45.192308
5,Mission,Euphoria Mourning,1999,56.626506
6,Wave Goodbye,Euphoria Mourning,1999,53.921569
7,Moonchild,Euphoria Mourning,1999,48.672566
8,Sweet Euphoria,Euphoria Mourning,1999,71.111111
9,Disappearing One,Euphoria Mourning,1999,39.583333


In [16]:
scatter_song_word_diversity_df.to_csv("scatter_song_word_diversity_df.csv")

In [17]:
song_word_diversity_df.to_csv("song_word_diversity_df.csv")

6. By how much did word diversity change from first album to last album for solo, Soundgarden, and Audioslave?

In [18]:
album_word_diversity_df = song_word_diversity_df.copy()

In [19]:
album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Euphoria Mourning', 'unique_text_count_processed_lyrics'].sum()

np.int64(703)

In [20]:
#calculate word diveristy mean for each album 
#use chart to show change 

unique_word_count_euphoriamourning = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Euphoria Mourning', 'unique_text_count_processed_lyrics'].sum()

unique_word_count_highertruth = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Higher Truth', 'unique_text_count_processed_lyrics'].sum()

unique_word_count_ultramegaok = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Ultramega OK', 'unique_text_count_processed_lyrics'].sum()

unique_word_count_kinganimal = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'King Animal', 'unique_text_count_processed_lyrics'].sum()

unique_word_count_audioslave = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Audioslave', 'unique_text_count_processed_lyrics'].sum()

unique_word_count_revelations = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Revelations', 'unique_text_count_processed_lyrics'].sum()

In [21]:
total_word_count_euphoriamourning = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Euphoria Mourning', 'word_count_processed_lyrics'].sum()

total_word_count_highertruth = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Higher Truth', 'word_count_processed_lyrics'].sum()

total_word_count_ultramegaok = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Ultramega OK', 'word_count_processed_lyrics'].sum()

total_word_count_kinganimal = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'King Animal', 'word_count_processed_lyrics'].sum()

total_word_count_audioslave = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Audioslave', 'word_count_processed_lyrics'].sum()

total_word_count_revelations = album_word_diversity_df.loc[album_word_diversity_df['album_name'] == 'Revelations', 'word_count_processed_lyrics'].sum()

In [22]:
word_diversity_score_euphoriamourning = unique_word_count_euphoriamourning/total_word_count_euphoriamourning*100

word_diversity_score_highertruth = unique_word_count_highertruth/total_word_count_highertruth*100

word_diversity_score_ultramegaok = unique_word_count_ultramegaok/total_word_count_ultramegaok*100

word_diversity_score_kinganimal = unique_word_count_kinganimal/total_word_count_kinganimal*100

word_diversity_score_audioslave = unique_word_count_audioslave/total_word_count_audioslave*100

word_diversity_score_revelations = unique_word_count_revelations/total_word_count_revelations*100

In [23]:
album_word_diversity_df=album_word_diversity_df.drop_duplicates('album_name')

In [24]:
album_word_diversity_df.drop(album_word_diversity_df[album_word_diversity_df["album_name"] == "Temple of the Dog"].index, inplace=True)

In [25]:
album_word_diversity_df.shape

(6, 8)

In [26]:
album_word_diversity_df['average_album_word_diversity']= [word_diversity_score_euphoriamourning,
                                                          word_diversity_score_highertruth,
                                                          word_diversity_score_ultramegaok,
                                                          word_diversity_score_kinganimal,
                                                          word_diversity_score_audioslave,
                                                          word_diversity_score_revelations]

In [27]:
album_word_diversity_df=album_word_diversity_df[['album_name','average_album_word_diversity']].sort_values('average_album_word_diversity',ascending = False)

In [28]:
album_word_diversity_df.reset_index(drop=True, inplace = True)

In [29]:
album_word_diversity_df

,album_name,average_album_word_diversity
0,Higher Truth,50.154936
1,Euphoria Mourning,45.827901
2,Ultramega OK,43.046358
3,Revelations,42.643229
4,Audioslave,42.118698
5,King Animal,41.580042


In [30]:
album_word_diversity_df.loc[0:2, "first_album"] = ['Euphoria Mourning', 'Ultramega OK', 'Audioslave']
album_word_diversity_df.loc[0:2, "last_album"] = ['Higher Truth', 'King Animal', 'Revelations']

album_word_diversity_df["start_value"] = album_word_diversity_df["first_album"].map(
    album_word_diversity_df.set_index("album_name")["average_album_word_diversity"]
)

album_word_diversity_df["end_value"] = album_word_diversity_df["last_album"].map(
    album_word_diversity_df.set_index("album_name")["average_album_word_diversity"]
)

In [31]:
range_chart_df = album_word_diversity_df.drop(columns = ['album_name', 'average_album_word_diversity']).copy()

In [32]:
range_chart_df = range_chart_df.drop(range(3, 6))

In [33]:
range_chart_df.to_csv("range_chart_df.csv")